File: app/src/main/java/danb/speedrunbrowser/api/PushNotificationData.java
package danb.speedrunbrowser.api;

import com.google.gson.Gson;

import java.util.Map;

import danb.speedrunbrowser.api.objects.Category;
import danb.speedrunbrowser.api.objects.Game;
import danb.speedrunbrowser.api.objects.LeaderboardRunEntry;
import danb.speedrunbrowser.api.objects.Level;

public class PushNotificationData {
    public LeaderboardRunEntry old_run;
    public LeaderboardRunEntry new_run;

    public Game game;
    public Category category;
    public Level level;

    public PushNotificationData(Map<String, String> data) {

        Gson gson = SpeedrunMiddlewareAPI.getGson();

        old_run = gson.fromJson(data.get("old_run"), LeaderboardRunEntry.class);
        new_run = gson.fromJson(data.get("new_run"), LeaderboardRunEntry.class);

        game = gson.fromJson(data.get("game"), Game.class);
        category = gson.fromJson(data.get("category"), Category.class);
        level = gson.fromJson(data.get("level"), Level.class);
    }
}

File: app/src/main/java/danb/speedrunbrowser/api/PushNotificationData.kt
package danb.speedrunbrowser.api

import com.google.gson.Gson

import danb.speedrunbrowser.api.objects.Category
import danb.speedrunbrowser.api.objects.Game
import danb.speedrunbrowser.api.objects.LeaderboardRunEntry
import danb.speedrunbrowser.api.objects.Level

class PushNotificationData(data: Map<String, String>) {
    val oldRun: LeaderboardRunEntry?
    val newRun: LeaderboardRunEntry

    val game: Game
    val category: Category
    val level: Level?

    init {

        val gson = SpeedrunMiddlewareAPI.gson

        oldRun = gson.fromJson(data["old_run"], LeaderboardRunEntry::class.java)
        newRun = gson.fromJson(data["new_run"], LeaderboardRunEntry::class.java)

        game = gson.fromJson(data["game"], Game::class.java)
        category = gson.fromJson(data["category"], Category::class.java)
        level = gson.fromJson(data["level"], Level::class.java)
    }
}
